esp/esp/program/modules/handlers/studentregphasezero.py

```diff
@@ -0,0 +1,220 @@
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public
+License along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+Contact information:
+MIT Educational Studies Program
+  84 Massachusetts Ave W20-467, Cambridge, MA 02139
+  Phone: 617-253-4882
+  Email: esp-webmasters@mit.edu
+Learning Unlimited, Inc.
+  527 Franklin St, Cambridge, MA 02139
+  Phone: 617-379-0178
+  Email: web-team@learningu.org
+"""
+
+from esp.utils.web import render_to_response
+from esp.middleware.threadlocalrequest import get_current_request
+from esp.program.modules.base import ProgramModuleObj, main_call, aux_call, meets_deadline, needs_student, meets_grade, meets_cap, no_auth, needs_admin
+from esp.users.models import Record, ESPUser, Permission
+from esp.program.models import PhaseZeroRecord
+from esp.program.modules.forms.phasezero import SubmitForm
+from esp.dbmail.models import send_mail
+from esp.tagdict.models import Tag
+
+from django.conf import settings
+from django.template.loader import render_to_string
+from django.contrib.auth.models import Group
+from django.db.models.query import Q
+
+import random, copy, datetime, re
+
+class StudentRegPhaseZero(ProgramModuleObj):
+    def students(self, QObject = False):
+        q_phasezero = Q(phasezerorecord__program=self.program)
+
+        if QObject:
+            retVal = {'phasezero': q_phasezero}
+            return retVal
+
+        retVal = {'phasezero': ESPUser.objects.filter(q_phasezero).distinct()}
+        return retVal
+
+    def studentDesc(self):
+        retVal = {'phasezero': """Students who have entered the Student Lottery."""}
+        return retVal
+
+    def isCompleted(self):
+        return get_current_request().user.canPassPhaseZero(self.program)
+
+    @classmethod
+    def module_properties(cls):
+        return {
+            "admin_title": "Student Registration Phase Zero",
+            "link_title": "Student Registration Phase Zero",
+            "module_type": "learn",
+            "seq": 2,
+            "required": True
+        }
+
+    @main_call
+    @needs_student
+    @meets_grade
+    def studentregphasezero(self, request, tl, one, two, module, extra, prog):
+        """
+        Serves the Phase Zero student reg page. The initial page includes a button
+        to enter the student lottery. Following entering the lottery, students are
+        served a confirmation page and lottery number form (for self-assigning groups).
+        """
+        context = {}
+        context['program'] = prog
+        user = request.user
+        in_lottery = PhaseZeroRecord.objects.filter(user=user, program=prog).exists()
+
+        op = ''
+        join_error = False
+        if 'op' in request.POST:
+            op = request.POST['op']
+        if op == 'join':
```

Comment:
Since this has totally separate code, can it just be a separate `@aux_call`?  Then you can get rid of `op`, and simplify some of the deeply-nested if blocks here.